tests/e2e_tests.c

```diff
@@ -49,48 +53,82 @@ static void	finally(int *out, int *save)
 	close(*save);
 }
 
+#ifdef FT_CALLOC_IF_TRAPPED
 static void	finally_err(int *out, int *save)
 {
 	fflush(stderr);
 	close(*out);
 	dup2(*save, fileno(stderr));
 	close(*save);
 }
+#endif
 
 static int	file_compare(char *exp_contens, char *act_fname)
 {
-	int fc = open(act_fname, O_RDONLY, 0600);
-	int len_exp = ft_strlen(exp_contens);
-	char *act = ft_calloc(sizeof(char), len_exp * 2 + 1);
+	#ifdef DEBUG
+	fprintf(stderr, "comparing files\n");
+	fflush(stderr);
+	fprintf(stderr, "expected [%p]\n", exp_contens);
+	fflush(stderr);
+	fprintf(stderr, "expected [%s]\n", exp_contens);
+	fflush(stderr);
+	fprintf(stderr, "actual fname [%s]\n", act_fname);
+	fflush(stderr);
+	#endif
+	FILE *f = fopen(act_fname, "rb");
+	fseek(f, 0, SEEK_END);
+	long fsize = ftell(f);
+	fseek(f, 0, SEEK_SET);
+	char *act = ft_calloc(sizeof(char), fsize + 1);
 	assert(act);
-	int len_act = read(fc, act, len_exp * 2);
-	assert(len_act - len_exp <= 3); // expected result doesn't store a number
-	int comp_res = ft_strncmp(exp_contens, act, ft_strlen(exp_contens) - 1);
+	fread(act, fsize, 1, f);
+	act[fsize] = 0;
+	fclose(f);
+
+	regex_t re;
+	char	*exp_re = ft_strjoin(exp_contens, "(Total ft_calloc_calls: [0-9]*\n)?");
+	assert(exp_re);
+	assert(regcomp(&re, exp_re, REG_EXTENDED | REG_NOSUB) == 0);
+	int comp_res = regexec(&re, act, 0, NULL, 0);
 	#ifdef DEBUG
-	ft_printf("comparison result %i, expected:[%s] (%i symbols)\nactual:[%s] (%i symbols)\n", comp_res, exp_contens, len_exp, act, len_act);
+	int	diff = -1;
+	for (int i = 0; i < fsize; i ++)
+		if (act[i] != exp_re[i])
+		{
+			diff = i;
+			break ;
+		}
+	ft_printf("comparison result %i, expected:{{%s}}\nactual:{{%s}}\nstrncmp %i\nstart to differ from index %i\n", comp_res, exp_re, act, ft_strncmp(exp_contens, act, fsize), diff);
+	fflush(stdout);
 	#endif
 	assert(comp_res == 0);
+	free(exp_re);
 	int i;
-	for (i = len_act - 2; act[i] >= '0' && act[i] <= '9'; i --);
-	char *sub = ft_substr(act, i, len_act);
+	for (i = fsize - 2; act[i] >= '0' && act[i] <= '9'; i --);
+	char *sub = ft_substr(act, i, fsize - i);
 	int mallocs = ft_atoi(sub, 0);
 	free(sub);
 	free(act);
-	close(fc);
 	return (mallocs);
 }
 
 static void	successful_execution(t_testcase *test, int *mallocs)
 {
 	int out, save;
-	system("(rm -r e2e_f && rm e2e.stdout) 2> /dev/null");
+	system("(rm -r e2e_f testf && rm e2e.stdout e2e.stderr) 2> /dev/null");
 	assert(system("mkdir e2e_f") == 0);
 	assert(system("cp minishell e2e_f/minishell") == 0);
 	catch("e2e.stdout", &out, &save);
 	char *tmp = ft_strjoin("./e2e_f/minishell ", test->cmd);
 	assert(!!tmp);
+	#ifdef DEBUG
+	fprintf(stderr, "executing [%s]\n", tmp);
+	#endif
 	assert(system(tmp) == 0);
 	finally(&out, &save);
+	#ifdef DEBUG
+	fprintf(stderr, "executing [%s]\n", tmp);
+	#endif
```

Comment:
i didn't find explanation in previous PR, anyway explain more detailed, what for lines 
```
if (ft_strcmp(key, "stdout") != 0)
			file_compare(((t_mapss_entry *)entry->content)->value, tmp);
```
i dont see test cases when key is not "stdout"

Reply:
This is for redirections. We don't have them implemented and tested yet.
For mock 3 there will be smth like
stdout: "" (empty)
out.txt: "1 11"

Reply:
ok , i didnt get it :) lets you explain when we meet